### PR TITLE
Upgrade Jetty to 9.4.6.v20170531

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
Verifier runs for both HTTP/1 and HTTP/2 look good.

### CODE REVIEW

File: http-server/src/main/java/io/airlift/http/server/HashLoginServiceProvider.java [hunk 1]
```diff
@@ -42,18 +37,9 @@ public HashLoginServiceProvider(HttpServerConfig config)
     public HashLoginService get()
```
Comment (nezihyigitbasi) [inline http-server/src/main/java/io/airlift/http/server/HashLoginServiceProvider.java]: `loadUsers()` method was throwing previously and it's removed in the new release, so nothing is throwing an `IOException` now. BTW that method was already a noop in 9.3.13.M0 so no ideas why we were calling it in the first place.

File: http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java [hunk 2]
```diff
@@ -289,4 +291,30 @@ public HttpClientConfig setKerberosRemoteServiceName(String serviceName)
         this.kerberosRemoteServiceName = serviceName;
         return this;
     }
+
+    public DataSize getHttp2InitialSessionReceiveWindowSize()
```
Comment (electrum) [inline http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java]: Add `@NotNull` and min/max. Thoughts on min/max values? Probably 1GB for max. Maybe 1kB for min?

File: http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java [hunk 3]
```diff
@@ -185,6 +186,8 @@ public JettyHttpClient(
         HttpClientTransport transport;
         if (config.isHttp2Enabled()) {
             HTTP2Client client = new HTTP2Client();
+            client.setInitialSessionRecvWindow(Ints.checkedCast(config.getHttp2InitialSessionReceiveWindowSize().toBytes()));
```
Comment (electrum) [inline http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java]: Use `toIntExact` with static import

File: http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java [hunk 4]
```diff
@@ -623,27 +626,31 @@ private static String dumpDestination(Destination destination)
...
-                .filter(exchange -> exchange != null)
+                .map(HttpConnectionOverHTTP.class::cast)
+                .map(connection -> connection.getHttpChannel().getHttpExchange())
+                .filter(Objects::nonNull)
                 .forEach(exchange -> requests.add(exchange.getRequest()));
 
-        return requests.stream().filter(request -> request != null).collect(Collectors.toList());
+        return requests.stream()
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
```
Comment (electrum) [inline http-client/src/main/java/io/airlift/http/client/jetty/JettyHttpClient.java]: Use `toImmutableList`

File: http-client/src/test/java/io/airlift/http/client/TestHttpClientConfig.java [hunk 5]
```diff
@@ -83,6 +85,8 @@ public void testExplicitPropertyMappings()
                 .put("http-client.authentication.enabled", "true")
                 .put("http-client.authentication.krb5.remote-service-name", "airlift")
                 .put("http-client.authentication.krb5.principal", "airlift-client")
+                .put("http-client.http2.session-receive-window-size", "8MB")
```
Comment (electrum) [inline http-client/src/test/java/io/airlift/http/client/TestHttpClientConfig.java]: Use a non power of two like 7MB that wouldn't be believable as a default.

File: http-server/src/main/java/io/airlift/http/server/HttpServer.java [hunk 6]
```diff
@@ -149,6 +146,8 @@ public HttpServer(HttpServerInfo httpServerInfo,
             Integer selectors = config.getHttpSelectorThreads();
             HttpConnectionFactory http1 = new HttpConnectionFactory(httpConfiguration);
             HTTP2CServerConnectionFactory http2c = new HTTP2CServerConnectionFactory(httpConfiguration);
+            http2c.setInitialSessionRecvWindow(Ints.checkedCast(config.getHttp2InitialSessionReceiveWindowSize().toBytes()));
```
Comment (electrum) [inline http-server/src/main/java/io/airlift/http/server/HttpServer.java]: `toIntExact`

File: http-server/src/main/java/io/airlift/http/server/HttpServer.java [hunk 7]
```diff
@@ -381,6 +380,10 @@ public void start()
             throws Exception
     {
         server.start();
+        //start() call already registers an error handler
```
Comment (electrum) [inline http-server/src/main/java/io/airlift/http/server/HttpServer.java]: How about
```java
// clear the error handler registered by start()

File: http-server/src/main/java/io/airlift/http/server/HttpServerConfig.java [hunk 8]
```diff
@@ -462,4 +464,30 @@ public HttpServerConfig setShowStackTrace(boolean showStackTrace)
         this.showStackTrace = showStackTrace;
         return this;
     }
+
+    public DataSize getHttp2InitialSessionReceiveWindowSize()
```
Comment (electrum) [inline http-server/src/main/java/io/airlift/http/server/HttpServerConfig.java]: Same comments as client config

File: http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java [hunk 9]
```diff
@@ -289,4 +291,30 @@ public HttpClientConfig setKerberosRemoteServiceName(String serviceName)
...
+    {
+        return http2InitialSessionReceiveWindowSize;
+    }
+
+    @Config("http-client.http2.session-receive-window-size")
+    @ConfigDescription("The initial size of session's flow control receive window for HTTP/2")
```
Comment (electrum) [inline http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java]: Nit: remove "The" so it starts with "Initial size of ..."

File: http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java (hunk 2 above)
Comment (nezihyigitbasi) [inline http-client/src/main/java/io/airlift/http/client/HttpClientConfig.java]: sounds reasonable.

File: http-server/src/main/java/io/airlift/http/server/HttpServer.java (hunk 6 above)
Comment (nezihyigitbasi) [inline http-server/src/main/java/io/airlift/http/server/HttpServer.java]: I will update, but we have a bunch of similar `checkedCast`s in this class.

**Comment by nezihyigitbasi** [comment]
@electrum updated.

**Comment by electrum** [comment]
I saw an error when running the tests on my laptop:
```
testErrorResponseBody
java.io.UncheckedIOException: java.nio.channels.ClosedChannelException
	at io.airlift.http.client.ResponseHandlerUtils.propagate(ResponseHandlerUtils.java:21)
	at io.airlift.http.client.StringResponseHandler.handleException(StringResponseHandler.java:50)
	at io.airlift.http.client.StringResponseHandler.handleException(StringResponseHandler.java:34)
	at io.airlift.http.client.jetty.JettyHttpClient.execute(JettyHttpClient.java:373)
	at io.airlift.http.client.jetty.TestJettyHttpClientSocksProxy.executeRequest(TestJettyHttpClientSocksProxy.java:55)
	at io.airlift.http.client.AbstractHttpClientTest.testErrorResponseBody(AbstractHttpClientTest.java:327)
Caused by: java.nio.channels.ClosedChannelException
	at org.eclipse.jetty.http2.HTTP2Session.onShutdown(HTTP2Session.java:875)
	at org.eclipse.jetty.http2.HTTP2Connection$HTTP2Producer.produce(HTTP2Connection.java:227)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:211)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:126)
	at org.eclipse.jetty.http2.HTTP2Connection.onFillable(HTTP2Connection.java:116)
	at org.eclipse.jetty.http2.HTTP2Connection$FillableCallback.succeeded(HTTP2Connection.java:254)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:110)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:135)
	at org.eclipse.jetty.util.thread.Invocable$InvocableExecutor.invoke(Invocable.java:222)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:294)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:199)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:673)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:591)
	at java.lang.Thread.run(Thread.java:745)
... Removed 28 stack frames

**Comment by electrum** [comment]
Merged, thanks!

**Pull request closed without merging**